azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java

```diff
@@ -105,6 +105,8 @@
   private final int maxConcurrentRunsOneFlow;
   QueuedExecutions queuedFlows;
   File cacheDir;
+  //make it immutable to ensure threadsafety
+  private volatile ImmutableSet<Executor> activeExecutors = null;
```

Comment:
How is the thread-safe guaranteed compared to HashSet?

Reply:
since immutable data structure is always thread safe.

Reply:
I'd like to learn which part of code need the thread-safe data structure, activeExecutors? Would you mind point it out?

Reply:
yes, so the class contains multiple pieces iterating over activeExecutor, and code piece populating activeExecutors from database. activeExecutors should be immune to race condition when iterating and populating activeExecutors are in different thread context.